### ISSUE TITLE
pytorch < 1.0.0 compatibility

### ISSUE THREAD

**Issue by ferrine** [ferrine/geoopt#32]
I have to add some weird code to support pytorch<1.0.0. Does it worth it at all?

Arguments against:
* It seems pytorch<1.0.0 is faster (why so?)
* Not all users have yet upgraded their code

Arguments for:
* It's a new package, research oriented audience

CC @soumith

**Comment by soumith** [comment]
could you give me a quick command / benchmark on where you se <1.0.0 to be faster?
We did have some regressions on advanced indexing, and indexing, I can check (with your benchmark) if there are other places we regressed and fix those as well.

**Comment by ferrine** [comment]
I have several benchmarks. First, I've spotted the issue in my lab work. I have regular ConvNets there, nothing special. This prevented me from putting efforts to support for pytorch 1.0. This was my GPU benchmark. I need some time to provide code to reproduce this ofc.

Second spotted issue is travis here. It runs pytorch-0.4.1, pytorch-1.0.0 , pytorch-nightly. This is my CPU menchmark. I see that despite tracing optimizations in `geoopt.optim` I have tests on pytorch-1.0.0/nightly to be much slower
https://travis-ci.com/ferrine/geoopt/builds/97819727
I also print tests durations in travis logs, so you can see and compare how long does it takes for the same test to run on torch-nightly and 0.4.1
torch-nightly https://travis-ci.com/ferrine/geoopt/jobs/171397573#L2823
torch-0.4.1 https://travis-ci.com/ferrine/geoopt/jobs/171397571#L1206
You see, ordering is about the same, but they all are slower for torch-nightly

You can expect the benchmarks in some during the week

**Comment by ferrine** [comment]
I can confirm, basic benchmarks (conv+batch norm) are reasonably faster using cuda90 or cpu. However, for some code in here I have divergent results. For cpu, I have faster retraction (involves gesv + advanced indexing assignments) using pytorch 0.4.1 especially for small matrix sizes (10).
`69.5 µs ± 121 ns -> 89.7 µs ± 221 ns`
This mostly correlates with tests speed. The difference for larger matrices is smaller and does not make much sense. Good news is that cuda has no such problem.
https://gist.github.com/ferrine/ac287080f8fc28d40b7823589fbd06cc

I think it might be reasonable to drop pytorch<1.0.0 support in the release and have it right before in 0.0.x series

**Comment by ferrine** [comment]
I've refactored code to speed up further development in #39
